hash_practice/exercises.py

```diff
@@ -22,8 +40,22 @@ def valid_sudoku(table):
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n2)
+        Space Complexity: O(1)
     """
-    pass
+    for row in table: 
+        for element in row: 
+            if element != ".":
+                if element not in "123456789":
+                    return False 
+                else: 
+                    if row.count(element) >1: 
+                        return False 
+                    if [row[i] for i in range(9)].count(element)>1:

```

Comment:
I know this problem is optional but I just want to note that the list comprehension `[row[i] for i in range(9)]` will not actually return the columns, but just a copy of `row` again (since all rows are 9 elements long).

To get the columns, you'll have to iterate over `table` not `row`.